senders/test_server_test.go

```diff
@@ -0,0 +1,73 @@
+package senders
+
+import (
+	"bufio"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"net/http"
+)
+
+func newTestServer(port int) *testServer {
+	metricLines := make([]string, 0)
+	return &testServer{
+		port:        port,
+		metricLines: metricLines,
+	}
+}
+
+type testServer struct {
+	port        int
+	metricLines []string
+	httpServer  *http.Server
+}
+
+func (server *testServer) start() {
+	addr := fmt.Sprintf(":%v", server.port)
+	handler := http.NewServeMux()
+	handler.HandleFunc("/", server.handler())
+	httpServer := &http.Server{Addr: addr, Handler: handler}
+
+	server.httpServer = httpServer
+	go func() {
+		if err := httpServer.ListenAndServe(); err != nil {
+			// handle err
+		}
+	}()
+}
+
+func (server *testServer) handler() func(writer http.ResponseWriter, request *http.Request) {
+	return func(writer http.ResponseWriter, request *http.Request) {
+		newLines, err := decodeMetricLines(request)
+		if err != nil {
+			writer.WriteHeader(500)
+		}
+		for _, metricLine := range newLines {
+			server.metricLines = append(server.metricLines, metricLine)
+		}
+		writer.WriteHeader(200)
+	}
+}
+
+func decodeMetricLines(request *http.Request) ([]string, error) {
+	metricLines := make([]string, 0)
+	reader, err := gzip.NewReader(request.Body)
+	if err != nil {
+		reader.Close()
+	}
+	scanner := bufio.NewScanner(reader)
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		metricLines = append(metricLines, line)
+	}
+
+	if err = reader.Close(); err != nil {
```

Comment:
Consider calling reader.Close() in a defer statement.  Once reader is successfully created, you have to close it when done anyway.

Reply:
This looked a little funny to me as well - it's modified from the gzip logic that is used to compress the lines in the reporter, so this pattern shows up there as well.

I think that `defer` would do the right thing in terms of making sure that the reader is always closed without having to call it in two different spots, but it would also mean that we couldn't return the `error` from `reader.Close`, which I think could cause us to drop some useful error messages, at least in theory. I've been inclined to keep using this pattern so that we're consistently returning all the error values we get.

I changed this up a little bit, but if you think there's a better way to use defer and still return the error, I'd definitely be up for trying it out.